File: python/complex-numbers/complex_numbers.py
from math import exp, cos, sin
import numbers


class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        if isinstance(other, ComplexNumber):
            return ComplexNumber(
                self.real + other.real, self.imaginary + other.imaginary
            )
        elif isinstance(other, numbers.Real):
            return ComplexNumber(self.real + other, self.imaginary)
        else:
            raise TypeError("Type not defined")

    def __mul__(self, other):
        if isinstance(other, ComplexNumber):
            return ComplexNumber(
                self.real * other.real - self.imaginary * other.imaginary,
                self.imaginary * other.real + self.real * other.imaginary,
            )
        elif isinstance(other, numbers.Real):
            return ComplexNumber(self.real * other, self.imaginary * other)
        else:
            raise TypeError("Type not defined")

    def __radd__(self, other):
        if isinstance(other, numbers.Real):
            return ComplexNumber(self.real + other, self.imaginary)
        else:
            raise TypeError("Type not defined")

    def __rmul__(self, other):
        if isinstance(other, numbers.Real):
            return ComplexNumber(self.real * other, self.imaginary * other)
        else:
            raise TypeError("Type not defined")

    def __sub__(self, other):
        if isinstance(other, ComplexNumber):
            return ComplexNumber(
                self.real - other.real, self.imaginary - other.imaginary
            )
        elif isinstance(other, numbers.Real):
            return ComplexNumber(self.real - other, self.imaginary)
        else:
            raise TypeError("Type not defined")

    def __float__(self):
        return float(self.real)

    def __repr__(self):
        return f"{self.real} + {self.imaginary}i"

    def __truediv__(self, other):
        if isinstance(other, ComplexNumber):
            denominator = float(other.conjugate() * other)
            numerator = other.conjugate() * self
            return ComplexNumber(
                numerator.real / denominator, numerator.imaginary / denominator
            )
        elif isinstance(other, numbers.Real):
            return ComplexNumber(self.real / other, self.imaginary / other)
        else:
            raise TypeError("Type not defined")

    def __abs__(self):
        return (self.real ** 2 + self.imaginary ** 2) ** 0.5

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        return ComplexNumber(
            exp(self.real) * cos(self.imaginary),
            exp(self.real) * sin(self.imaginary),
        )

File: julia/complex-numbers/complex-numbers.jl
import Base:+,-,*,/,≈,==,exp, sqrt,show
struct ComplexNumber <: Number
    real::Float64
    imaginary::Float64
end
+(a::Number, b::ComplexNumber) = b + a
+(a::ComplexNumber, b::Number)  = ComplexNumber(a.real + b, a.imaginary)
+(a::ComplexNumber,b::ComplexNumber) =  ComplexNumber(a.real + b.real,a.imaginary + b.imaginary)
-(a::ComplexNumber,b::ComplexNumber) =  ComplexNumber(a.real  - b.real,a.imaginary - b.imaginary)
*(a::Number,b::ComplexNumber) =  b * a
*(a::ComplexNumber,b::Number) =  ComplexNumber(a.real * b,a.imaginary * b)
*(a::ComplexNumber,b::ComplexNumber) =  ComplexNumber(a.real * b.real  - a.imaginary * b.imaginary,a.real*b.imaginary + a.imaginary * b.real)
conj(n::ComplexNumber) = ComplexNumber(n.real,n.imaginary  == 0 ? 0 : -n.imaginary)
≈(a::ComplexNumber,b::ComplexNumber) =  isapprox(a.real,b.real,atol=1e-6)  && isapprox(a.imaginary,b.imaginary,atol=1e-6)
==(a::ComplexNumber,b::ComplexNumber) = a.real == b.real && a.imaginary == b.imaginary 
/(a::ComplexNumber,b::Number) =  ComplexNumber(a.real/b, a.imaginary/b)
/(a::ComplexNumber,b::ComplexNumber) =  conj(b) * a / (conj(b) * b).real
abs(n::ComplexNumber) = sqrt(n.real^2+n.imaginary^2)
real(n::ComplexNumber) =  n.real
imag(n::ComplexNumber) = n.imaginary
exp(n::ComplexNumber) = ComplexNumber(exp(n.real) * cos(n.imaginary), exp(n.real) * sin(n.imaginary))
#define jm
jm = ComplexNumber(0,1)